chacha20_poly1305/src/poly1305.rs

```diff
@@ -0,0 +1,274 @@
+        for i in 1..4 {
+            self.acc[i + 1] += self.acc[i] >> CARRY;
+        }
+        self.acc[0] += (self.acc[4] >> CARRY) * 5;
+        self.acc[1] += self.acc[0] >> CARRY;
+        for i in 0..self.acc.len() {
+            self.acc[i] &= BITMASK;
+        }
+        // Reduce.
+        let mut t = self.acc;
+        t[0] += 5;
+        t[4] = t[4].wrapping_sub(1 << CARRY);
+        for i in 0..3 {
+            t[i + 1] += t[i] >> CARRY;
+        }
+        t[4] = t[4].wrapping_add(t[3] >> CARRY);
+        for t in t.iter_mut().take(4) {
+            *t &= BITMASK;
+        }
+        // Convert acc to a 4 item array.
+        let mask = (t[4] >> 31).wrapping_sub(1);
+        for (i, t) in t.iter().enumerate().take(self.acc.len()) {
+            self.acc[i] = t & mask | self.acc[i] & !mask;
+        }
+        // Voodoo from donna to convert to [u32; 4].
+        let a0 = self.acc[0] | self.acc[1] << 26;
+        let a1 = self.acc[1] >> 6 | self.acc[2] << 20;
+        let a2 = self.acc[2] >> 12 | self.acc[3] << 14;
+        let a3 = self.acc[3] >> 18 | self.acc[4] << 8;
+        let a = [a0, a1, a2, a3];
+        // a + s
+        let mut tag: [u64; 4] = [0; 4];
+        for i in 0..4 {
+            tag[i] = a[i] as u64 + self.s[i] as u64;
+        }
+
+        // Carry.
+        for i in 0..3 {
+            tag[i + 1] += tag[i] >> 32;
+        }
+
+        // Return the 16 least significant bytes.
+        let mut ret: [u8; 16] = [0; 16];
+        for i in 0..tag.len() {
+            let bytes = (tag[i] as u32).to_le_bytes();
+            ret[i * 4..(i + 1) * 4].copy_from_slice(&bytes);
+        }
+        ret
+    }
+
+    fn r_times_a(&mut self) {
+        // Multiply and reduce.
+        // While this looks complicated, it is a variation of schoolbook multiplication,
+        // described well in an article here: https://loup-vaillant.fr/tutorials/poly1305-design
+        let mut t = [0; 5];
+        for i in 0..5 {
+            for (j, t) in t.iter_mut().enumerate() {
+                let modulus: u64 = if i > j { 5 } else { 1 };
+                let start = (5 - i) % 5;
+                *t += modulus * self.r[i] as u64 * self.acc[(start + j) % 5] as u64;
+            }
+        }
+        // Carry.
+        for i in 0..4 {
+            t[i + 1] += t[i] >> CARRY;
+        }
+        // Mask.
+        for (i, t) in t.iter().enumerate().take(self.acc.len()) {
+            self.acc[i] = *t as u32 & BITMASK;
+        }
+        // Carry and mask first limb.
+        self.acc[0] += (t[4] >> CARRY) as u32 * 5;
+        self.acc[1] += self.acc[0] >> CARRY;
+        self.acc[0] &= BITMASK;
+    }
+}
+
+// Encode 16-byte (tag sized), unless is_last flag set to true, piece of message into 5 26-bit limbs.
+fn prepare_padded_message_slice(msg: &[u8], is_last: bool) -> [u32; 5] {
+    let hi_bit: u32 = if is_last { 0 } else { 1 << 24 };
+    let mut fmt_msg = [0u8; 17];
+    fmt_msg[..msg.len()].clone_from_slice(msg);
```

Comment:
Why `clone` and not `copy`?

Reply:
Can't think of a good reason, switched it to `copy`.